### PR TITLE
Fixes and enhs for the visual dicom browser

### PR DESCRIPTION
**Pull request by Punzo**
I have adressed all items from:

Settings : https://github.com/commontk/CTK/issues/1162#Settings
Patient Selection: https://github.com/commontk/CTK/issues/1162#Patientselection

and:

last of Logic: https://github.com/commontk/CTK/issues/1162#Logic
second of Filtering: https://github.com/commontk/CTK/issues/1162#Filtering

See also https://github.com/Slicer/Slicer/pull/7676

### CODE REVIEW

File: Libs/DICOM/Core/ctkDICOMEchoWorker.h
Comment (jcfr) [inline Libs/DICOM/Core/ctkDICOMEchoWorker.h:53]: Add doxygen grouping `///@{` ...

File: Libs/DICOM/Core/ctkDICOMQuery.cpp
Comment (jcfr) [inline Libs/DICOM/Core/ctkDICOMQuery.cpp:176]: indent

File: Libs/DICOM/Core/ctkDICOMQuery.cpp
Comment (jcfr) [inline Libs/DICOM/Core/ctkDICOMQuery.cpp:180]: Use `QMuterLocker`

**Review by jcfr — COMMENTED** [review]
Minor stylistic changes. Have not tested yet.

File: Libs/Core/ctkJobScheduler.cpp
Comment (Punzo) [reply Libs/Core/ctkJobScheduler.cpp:228]: I first tried to use the QMutexLocker as Qt suggest (less error prone, cleaner implementation, etc...). However I had really bad performance issues when a method with QMutexLocker  calls another with also a QMutexLocker. I don't know why, but using the lock/unlock mechanism I did not have the slow-downs issues. I will try to investigate more.

File: Libs/Core/ctkJobScheduler.cpp
Comment (Punzo) [reply Libs/Core/ctkJobScheduler.cpp:30]: ok

File: Libs/DICOM/Core/ctkDICOMDatabase_p.h
Comment (Punzo) [reply Libs/DICOM/Core/ctkDICOMDatabase_p.h:196]: no in this case it should be the databaseConnectionNameID. 

```
bool insertPatient(const ctkDICOMItem& dataset,
    const QString& patientID,
    const QString& patientsName,
    int& databasePatientID);
```
    
```
bool insertConnectionName(int& databaseConnectionNameID,
    int databasePatientID,
    const QString& connectionName);
```

`databaseConnectionNameID ` in `insertConnectionName `is the equivalent of `databasePatientID`  in `insertPatient`

File: Libs/DICOM/Core/ctkDICOMEcho.cpp
Comment (Punzo) [reply Libs/DICOM/Core/ctkDICOMEcho.cpp:47]: ok

File: Libs/DICOM/Core/ctkDICOMEcho.cpp
Comment (Punzo) [reply Libs/DICOM/Core/ctkDICOMEcho.cpp:102]: ok

File: Libs/DICOM/Core/ctkDICOMEcho.h
Comment (Punzo) [reply Libs/DICOM/Core/ctkDICOMEcho.h:97]: indeed, thanks for catching it (:

File: Libs/DICOM/Core/ctkDICOMEchoWorker.h
Comment (Punzo) [reply Libs/DICOM/Core/ctkDICOMEchoWorker.h:53]: ok

File: Libs/DICOM/Core/ctkDICOMQuery.cpp
Comment (Punzo) [reply Libs/DICOM/Core/ctkDICOMQuery.cpp:176]: ok

File: Libs/Core/ctkJobScheduler.cpp
Comment (Punzo) [reply Libs/Core/ctkJobScheduler.cpp:239]: See https://github.com/commontk/CTK/pull/1201#discussion_r1553202391

File: Libs/Core/ctkJobScheduler.cpp
Comment (Punzo) [reply Libs/Core/ctkJobScheduler.cpp:338]: see https://github.com/commontk/CTK/pull/1201#discussion_r1553202391

File: Libs/DICOM/Core/ctkDICOMQuery.cpp
Comment (Punzo) [reply Libs/DICOM/Core/ctkDICOMQuery.cpp:180]: see https://github.com/commontk/CTK/pull/1201#discussion_r1553202391

**Comment by Punzo** [comment]
> Minor stylistic changes. Have not tested yet.

thanks Jc for the review. I have replied at the comments. I will apply them on Monday. 

If you will have the chance to test any further feedback will be very welcome :smiley:

Also guys @jcfr @lassoan @pieper  please note that commits https://github.com/commontk/CTK/pull/1201/commits/c8ac922669f3f60114e27f27e670dbc120504fc4
https://github.com/commontk/CTK/pull/1201/commits/e4093e28c56e864cee0fa6b10ad7ac87b6f1f244

 introduce a new table in the database, so I have updated the database schema version.
 
 and this commit https://github.com/commontk/CTK/pull/1201/commits/cdda814f4c01469c3a452b71db3e165d6c6a1d7e

fix an issue that I found in the dicomDatabse, where patient metadata were inserted either with empty patientName or patientID. This created a lot of perfoamnces issues because in many methods is not actually allowed having either one of them empty (e.g. compositeID, etc....).

File: Libs/Core/ctkJobScheduler.cpp
Comment (lassoan) [reply Libs/Core/ctkJobScheduler.cpp:228]: Maybe you need to reduce the locked scope  by adding putting the QMutexLocker in a `{}` block within the method. You can also do step-by-step debugging into the QMutexLocker to see if it does anything else than calling `lock()` and `unlock`.

**Comment by lassoan** [comment]
> Also guys @jcfr @lassoan @pieper please note that commits [c8ac922](https://github.com/commontk/CTK/commit/c8ac922669f3f60114e27f27e670dbc120504fc4) [e4093e2](https://github.com/commontk/CTK/commit/e4093e28c56e864cee0fa6b10ad7ac87b6f1f244)
> 
> introduce a new table in the database, so I have updated the database schema version.

Adding a new table just for storing multiple connection names for a patient is not worth it. It means extra complication when creating or deleting a patient (you need to create/remove extra records in another table) and even with this extra table, it is very limited and rigid what we can store. Instead, we could add a single new column in the patient table to store the list of of allowed and denied servers (need to store the list of denied servers, too, so that we don't need to keep bugging the user asking about a server that the user previously already told not to use for that patient). The format of the field could be json, as we already use it (for specifying display column format). The new column name could be `Connections` and could store something like this: `{"allow":["pacs1","pacs2"],"deny":["pacs5"]}`. Query/retrieve (and DICOM in general) is object-oriented anyway (e.g., you need to query each patient from the server), so it is very natural to use this kind of object-oriented interface, and this connections configuration data is small, therefore the performance should be as good (or even better) than with a relational database approach that uses a separate table.

File: Libs/Core/ctkJobScheduler.cpp
Comment (Punzo) [reply Libs/Core/ctkJobScheduler.cpp:228]: thanks for the suggestion, I will try to investigate it further next week

**Comment by Punzo** [comment]
> > Also guys @jcfr @lassoan @pieper please note that commits [c8ac922](https://github.com/commontk/CTK/commit/c8ac922669f3f60114e27f27e670dbc120504fc4) [e4093e2](https://github.com/commontk/CTK/commit/e4093e28c56e864cee0fa6b10ad7ac87b6f1f244)
> > introduce a new table in the database, so I have updated the database schema version.
> 
> Adding a new table just for storing multiple connection names for a patient is not worth it. It means extra complication when creating or deleting a patient (you need to create/remove extra records in another table) and even with this extra table, it is very limited and rigid what we can store. Instead, we could add a single new column in the patient table to store the list of allowed and denied servers (need to store the list of denied servers, too, so that we don't need to keep bugging the user asking about a server that the user previously already told not to use for that patient). Query/retrieve (and DICOM in general) is object-oriented anyway (e.g., you need to query each patient from the server), so it is very natural to use this kind of object-oriented interface, and this connections configuration data is small, therefore the performance should be as good (or even better) than with a relational database approach that uses a separate table.
> 
> The format of the field could be json, as we already use it (for specifying display column format). The new column name could be `Connections` and could store something like this: `{"allow":["pacs1","pacs2"],"deny":["pacs5"]}`. If you don't think we need a deny list for the current GUI implementation then we can just use `{"allow":["pacs1","pacs2"]}` for now. Using json is still worth it as it is more future-proof (if we need to add more per-patient connection configuration data then it can all go there), the syntax is self-explaining (no need to document what separator character is used, how to escape special characters, etc.), and we don't need to implement a parser.

I agree, I had the issue of the security warning been raised too many times. I will apply your proposed design next week.

File: Libs/Core/ctkJobScheduler.cpp
Comment (Punzo) [reply Libs/Core/ctkJobScheduler.cpp:228]: > Maybe you need to reduce the locked scope by adding putting the QMutexLocker in a `{}` block within the method.

thanks for the advice, it was indeed the scope. Using the QMutexLocker in {} blocks fixed the issue.

**Comment by Punzo** [comment]
> thanks Jc for the review. I have replied at the comments. I will apply them on Monda

@jcfr review applied in https://github.com/commontk/CTK/pull/1201/commits/8a61e8acf2a867ee334cfc5992ac0ce7cc6fcb22

**Comment by Punzo** [comment]
> I agree, I had the issue of the security warning been raised too many times. I will apply your proposed design next week.

@lassoan done in https://github.com/commontk/CTK/pull/1201/commits/dca47c9259261d5664621390cbca3a5029c3f250

File: Libs/DICOM/Widgets/ctkDICOMPatientItemWidget.cpp
Comment (lassoan) [inline Libs/DICOM/Widgets/ctkDICOMPatientItemWidget.cpp:463]: It would be better not to introduce "Secure". We are just talking about allowed and denied servers, allowed does it mean a server is secure; denied does not mean that the server is not secure.
```suggestion
bool ctkDICOMPatientItemWidgetPrivate::areAllActiveServersAllowed()
```

File: Libs/DICOM/Widgets/ctkDICOMPatientItemWidget.cpp
Comment (lassoan) [inline Libs/DICOM/Widgets/ctkDICOMPatientItemWidget.cpp:480]:
```suggestion
void ctkDICOMPatientItemWidgetPrivate::setAllDeniedServerEnabledStatus(bool enabled)
```

File: Libs/DICOM/Widgets/ctkDICOMPatientItemWidget.cpp
Comment (lassoan) [inline Libs/DICOM/Widgets/ctkDICOMPatientItemWidget.cpp:501]: not security - remove this from everywhere else (in the future maybe we'll introduce the concept of secure or insecure servers, but allow/deny is not that)

File: Libs/DICOM/Widgets/ctkDICOMPatientItemWidget.cpp
Comment (lassoan) [inline Libs/DICOM/Widgets/ctkDICOMPatientItemWidget.cpp:508]: Instead of q->tr, you always need to specify the exact translation context (Qt linguist translatable string extraction tool just performs simple code parsing, cannot determine the class name from q)

```suggestion
    QString warningString = ctkDICOMPatientItemWidget::tr("One or more active servers have been marked with an unrecognized security level. \n"
```

File: Libs/DICOM/Widgets/ctkDICOMPatientItemWidget.cpp
Comment (lassoan) [inline Libs/DICOM/Widgets/ctkDICOMPatientItemWidget.cpp:522]: replace q->tr everywhere

File: Libs/DICOM/Widgets/ctkDICOMPatientItemWidget.cpp
Comment (lassoan) [inline Libs/DICOM/Widgets/ctkDICOMPatientItemWidget.cpp:729]: Always use "allowed" and "denied" (not enabled/disabled, not secure/insecure)

File: Libs/DICOM/Widgets/ctkDICOMPatientItemWidget.cpp
Comment (lassoan) [inline Libs/DICOM/Widgets/ctkDICOMPatientItemWidget.cpp:1031]: Qt does not use `get` for accessors and uses `count` instead of `numberOf`
```suggestion
  if (queryRetrieve && d->Scheduler && d->Scheduler->queryRetrieveServersCount() > 0)
```

File: Libs/DICOM/Widgets/ctkDICOMPatientItemWidget.h
Comment (lassoan) [inline Libs/DICOM/Widgets/ctkDICOMPatientItemWidget.h:166]:
```suggestion
  Q_INVOKABLE ctkDICOMStudyItemWidget* geStudyItemWidgetByStudyItem(const QString& studyItem);
```
```suggestion
  Q_INVOKABLE ctkDICOMStudyItemWidget* studyItemWidgetByStudyItem(const QString& studyItem);
```

File: Libs/DICOM/Widgets/ctkDICOMPatientItemWidget.h
Comment (lassoan) [inline Libs/DICOM/Widgets/ctkDICOMPatientItemWidget.h:174]: Can we avoid this special `All` value? I would not allow all servers by default anyway (just where the image is retrieved from). It is also quite messy - the string needs to be translated, someone might accidentally use `All` as connection name, etc. (a separate flag or enum would be cleaner, but we can add it in the future, if we find that it is actually necessary to always automatically allow querying patient data from a server that did not originally provide it).

File: Libs/DICOM/Widgets/ctkDICOMVisualBrowserWidget.cpp
Comment (lassoan) [inline Libs/DICOM/Widgets/ctkDICOMVisualBrowserWidget.cpp:1123]:
```suggestion
  return patientIndex;
```

**Review by lassoan — COMMENTED** [review]
Really impressive work Davide, lots of fixes and improvements. I've added a number of comments they are mostly all trivial.

**Comment by Punzo** [comment]
> Really impressive work Davide, lots of fixes and improvements. I've added a number of comments they are mostly all trivial.

thanks and thanks for the review. I will apply it asap.

**Comment by Punzo** [comment]
> thanks and thanks for the review. I will apply it asap.
done!

**Comment by lassoan** [comment]
Let us know when this is ready for review.

**Comment by Punzo** [comment]
> Let us know when this is ready for review.

Ok, I'm planning to do a last round of manual testing tomorrow, in addition to the automated tests. I'll ping you when I have finished.

**Comment by Punzo** [comment]
@lassoan tested with https://github.com/Slicer/Slicer/pull/7676. This PR should be ready, please let me know if you have any other feedback.

**Comment by lassoan** [comment]
It would be nice to fix these before merging:

- In dark mode, some fields are not visible at all (see Settings section):

image

- I don't understand this sentence: "One or more active servers have been marked with an unrecognized access level. A server is marked as such if it deviates from the cached source of patient data server stored in the local DICOM database." Since we have a good default value ("trusted" flag), and popups are annoying, I don't think we need the popup at all. If query of the patient from the server was not explicitly disabled or enabled then decide based on the "trusted" flag of the server.
- Make checkboxes in "Servers" list in patient section tri-state. If grayed out then it means that the default is used (trusted flag).
- It is odd that "Servers" list in patient section is in a collapsible section. Why not a simple label+widget?
- Query/Retrieve checkbox of the search icon is not clear. "Query/Retrieve from servers" would be more clear and the tooltip could mention servers, too.

**Comment by Punzo** [comment]
> It would be nice to fix these before merging:
> 
> * In dark mode, some fields are not visible at all (see Settings section):
> 
> image
> 
> * I don't understand this sentence: "One or more active servers have been marked with an unrecognized access level. A server is marked as such if it deviates from the cached source of patient data server stored in the local DICOM database." Since we have a good default value ("trusted" flag), and popups are annoying, I don't think we need the popup at all. If query of the patient from the server was not explicitly disabled or enabled then decide based on the "trusted" flag of the server.
> * Make checkboxes in "Servers" list in patient section tri-state. If grayed out then it means that the default is used (trusted flag).
> * It is odd that "Servers" list in patient section is in a collapsible section. Why not a simple label+widget?
> * Query/Retrieve checkbox of the search icon is not clear. "Query/Retrieve from servers" would be more clear and the tooltip could mention servers, too.

1) Regarding the color in dark mode, I have fixed it. But we should really move await from palette and use only the stylesheet approach in all the CTK classes (i.e. https://github.com/commontk/CTK/issues/1162#UIcustomization). I can tackle it, we should also add @sjh26 in this loop.
2) I agree on the trusted servers behaviour/logic, I have applied it.

Done! 

Tomorrow I will do another set of manual tests. I will let you know when I have finished.

**Comment by Punzo** [comment]
> 1. Regarding the color in dark mode, I have fixed it. But we should really move await from palette and use only the stylesheet approach in all the CTK classes (i.e. [Roadmap list of issues/enhancements for Visual DICOM Browser #1162 (comment)](https://github.com/commontk/CTK/issues/1162#UIcustomization)). I can tackle it, we should also add @sjh26 in this loop.

Ideally the solution should:
1) use one unique style file containing the stylesheet for all the CTK classes
2) change color of qt widgets in CTK dynamically with the Property Selector, e.g.: https://wiki.qt.io/Dynamic_Properties_and_Stylesheets
3) custom Slicer app can chnage the style, simply rewriting and reloading the style file

**Comment by Punzo** [comment]
> 1. Regarding the color in dark mode, I have fixed it. But we should really move await from palette and use only the stylesheet approach in all the CTK classes (i.e. [Roadmap list of issues/enhancements for Visual DICOM Browser #1162 (comment)](https://github.com/commontk/CTK/issues/1162#UIcustomization)). I can tackle it, we should also add @sjh26 in this loop.

Ideally the solution should:
1) use one unique style file containing the stylesheet for all the CTK classes
2) change color of qt widgets in CTK dynamically with the Property Selector, e.g.: https://wiki.qt.io/Dynamic_Properties_and_Stylesheets
3) custom Slicer apps should be able to change the style by simply rewriting and reloading the style file

This is going to be addressed in another PR, since this PR is already too big. @sjh26, @lassoan @jcfr @pieper please let me know if you have any feedback/advice on this before I start any implementation. If necessary we can also do a videocall to discuss it.

**Comment by Punzo** [comment]
@lassoan I have done. For me it is good to go, please let me know if you have any further feedback.

**Comment by Punzo** [comment]
> > 1. Regarding the color in dark mode, I have fixed it. But we should really move await from palette and use only the stylesheet approach in all the CTK classes (i.e. [Roadmap list of issues/enhancements for Visual DICOM Browser #1162 (comment)](https://github.com/commontk/CTK/issues/1162#UIcustomization)). I can tackle it, we should also add @sjh26 in this loop.
> 
> I still have to review all CTK classes that uses the palette approach, but ideally the solution should:
> 
> 1. use one unique style file containing the stylesheet for all the CTK classes
> 2. change color of qt widgets in CTK dynamically with the Property Selector, e.g.: https://wiki.qt.io/Dynamic_Properties_and_Stylesheets
> 3. Slicer light/dark mode should also be set by a Property Selector (at the main window level? not sure). The style file then should define dynamic colors for the CTK classes for both modes.
> 4. custom Slicer apps should be able to change the style by simply rewriting and reloading the style file
> 
> This is going to be addressed in another PR, since this PR is already too big. @sjh26, @lassoan @jcfr @pieper please let me know if you have any feedback/advice on this before I start any implementation. If necessary we can also do a videocall to discuss it.

Provided a barebones example in https://github.com/Punzo/CTK/pull/4

**Review by lassoan — APPROVED** [review]
Thank you, I've tested on Windows and everything worked nicely.

**Comment by Punzo** [comment]
> Thank you, I've tested on Windows and everything worked nicely.

Thanks for testing and merging!

**Pull request merged**
